lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -544,6 +544,183 @@ static void mergeMatMul(Function *F) {
+
+    // For each slice:
+    for (SliceNode *SN : input) {
+      // Ignore slices of invalid types.
+      if (lastSlice->getResult().getType() != SN->getResult().getType()) {
+        continue;
+      }
+
+      // Check if SN comes after the last slice in the sequence.
+      if (areSlicesConsecutive(lastSlice, SN, dimension)) {
+        // Add the consecutive slice and schedule another iteration.
+        lastSlice = SN;
+        order.push_back(lastSlice);
+        addedSlice = true;
+        continue;
+      }
+    }
+  } // While adding new slices.
+
+  // Check that the last slice completes the tensor.
+  auto startCoor = lastSlice->getStart();
+  auto resDim = lastSlice->getResult()->getType()->dims();
+  auto inDim = lastSlice->getInput()->getType()->dims();
+
+  // Check if for all dimensions, the size of the result tensor plus the start
+  // coordinate matches the size of the tensor.
+  for (int i = 0, e = startCoor.size(); i < e; i++) {
+    if (startCoor[i] + resDim[i] != inDim[i])
+      return false;
+  }
+
+  // Report success if we found at least two slices that extract from the input.
+  return order.size() > 1;
+}
+
+/// Merge multiple batched add nodes into a large batched-add node.
+static void mergeBatchedAdd(Function *F) {
+  auto &nodes = F->getNodes();
+
+  // We index the batched add nodes by the slice operand.
+  llvm::DenseMap<Node *, std::vector<BatchedAddNode *>> rightBAUsers;
+
+  // Collect all of the batched add nodes and index them by the 'slice' operand.
+  for (auto const &node : nodes) {
+    if (auto *BA = dyn_cast<BatchedAddNode>(node)) {
+      rightBAUsers[BA->getSlice().getNode()].push_back(BA);
+    }
+  }
+
+  // For each 'slice' that batched add nodes access:
+  for (auto &it : rightBAUsers) {
+    auto &BAs = it.second;
+
+    // Collects the left-hand-side operands that the batched-adds add into. We
+    // only collect 'slice' nodes.
+    std::vector<SliceNode *> slices;
+
+    for (auto *BA : BAs) {
+      if (auto *S = dyn_cast<SliceNode>(BA->getBatch().getNode())) {
+        slices.push_back(S);
+      }
+    }
+
+    // Check if the slice nodes that we've collected cover a whole tensor.
+    std::vector<SliceNode *> order;
+    bool found = orderConsecutiveSlices(slices, 0, order);
+
+    if (!found) {
+      continue;
+    }
+
+    // We found a sequence of batched-add-slice that cover the input tensor. We
+    // can transform the graph and create one big batched-add.
+    std::vector<Node *> newSlices;
+    SliceNode *S = llvm::cast<SliceNode>(order[0]);
+    auto *BA = F->createBatchedAdd("mergedBA", S->getInput(), it.first);
+
+    // Create the new slices.
+    for (int i = 0, e = order.size(); i < e; i++) {
+      auto *orig = order[i];
+      newSlices.push_back(F->createSlice(orig->getName(), BA, orig->getStart(),
+                                         orig->getType()));
+    }
+
+    // Update all of our batched add nodes to use a slice from the vectorized
+    // batched add.
```

Comment:
nit: This comment is a bit confusing -- If I'm understanding correctly, might be better to say something like: "Replace the original individual batched adds with corresponding slices from the new merged batch add."